lib/logdna/client.rb

```diff
@@ -44,95 +62,130 @@ def process_message(msg, opts = {})
         e: "ls",
-        ls: @buffer
+        ls: slice
       }.to_json
 
-      handle_exception = lambda do |message|
-        @internal_logger.debug(message)
-        @exception_flag = true
-        @side_message_lock.synchronize do
-          @side_messages.concat(@buffer)
-        end
-      end
+      flush_id = "#{Random.uuid} [#{slice.length} lines]"
+      error_header = "Flush {#{flush_id}} failed."
+      tries = 0
+      loop do
+        tries += 1
 
-      begin
-        @response = Net::HTTP.start(
-          @uri.hostname,
-          @uri.port,
-          use_ssl: @uri.scheme == "https"
-        ) do |http|
-          http.request(@request)
+        if tries >= 4
+          @internal_logger.debug("Flush {#{flush_id}} exceeded 3 tries. Discarding flush buffer")
+          break
         end
-        if @response.is_a?(Net::HTTPForbidden)
-          @internal_logger.debug("Please provide a valid ingestion key")
-        elsif !@response.is_a?(Net::HTTPSuccess)
-          handle_exception.call("The response is not successful #{@response}")
-        end
-        @exception_flag = false
-      rescue SocketError
-        handle_exception.call("Network connectivity issue")
-      rescue Errno::ECONNREFUSED => e
-        handle_exception.call("The server is down. #{e.message}")
-      rescue Timeout::Error => e
-        handle_exception.call("Timeout error occurred. #{e.message}")
-      ensure
-        @buffer.clear
-      end
-    end
 
-    def flush
-      if @lock.try_lock
-        @flush_scheduled = false
-        if @buffer.any? || @side_messages.any?
-          send_request
+        begin
+          response = Net::HTTP.start(
+            @uri.hostname,
+            @uri.port,
+            use_ssl: @uri.scheme == "https"
+          ) do |http|
+            http.request(@request)
+          end
+
+          code = response.code.to_i
+          if code == 401 || code == 403
+            @internal_logger.debug("#{error_header} Please provide a valid ingestion key. Discarding flush buffer")
+            break
+          elsif code == 408 || code == 500 || code == 504
+            # These codes might indicate a temporary ingester issue
+            @internal_logger.debug("#{error_header} The request failed #{response}. Retrying")
+          elsif code == 200
+            break
+          else
+            @internal_logger.debug("#{error_header} The request failed #{response}. Discarding flush buffer")
+            break
+          end
+        rescue SocketError
+          @internal_logger.debug("#{error_header} Network connectivity issue. Retrying")
+        rescue Errno::ECONNREFUSED => e
+          @internal_logger.debug("#{error_header} The server is down. #{e.message}. Retrying")
+        rescue Timeout::Error => e
+          @internal_logger.debug("#{error_header} Timeout error occurred. #{e.message}. Retrying")
         end
-        @lock.unlock
-      else
-        schedule_flush
       end
     end
 
     def exitout
```

Comment:
This isn't called any more, is it?

https://github.com/logdna/ruby/pull/32/files#diff-2b8c8f0622fe34d1ba675085e0f713592ec2ff702622e8b17f1577cb682f442dL130

Reply:
This is still called when the main logger class' `close` is called